lib/src/cli/flutter_cli.dart

```diff
@@ -65,55 +62,182 @@ class Flutter {
   static Future<void> test({
     String cwd = '.',
     bool recursive = false,
-    void Function([String?]) Function(String message)? progress,
-  }) async {
-    await _runCommand(
-      cmd: (cwd) async {
-        final installDone = progress?.call(
-          'Running "flutter test" in $cwd',
+    void Function(String)? stdout,
+    void Function(String)? stderr,
+  }) {
+    return _runCommand(
+      cmd: (cwd) {
+        void noop(String? _) {}
+        stdout?.call('Running "flutter test" in $cwd...\n');
+        return _flutterTest(
+          cwd: cwd,
+          stdout: stdout ?? noop,
+          stderr: stderr ?? noop,
         );
-        try {
-          final result = await _Cmd.run(
-            'flutter',
-            ['test'],
-            workingDirectory: cwd,
-          );
-          return result;
-        } catch (_) {
-          rethrow;
-        } finally {
-          installDone?.call();
-        }
       },
       cwd: cwd,
       recursive: recursive,
     );
   }
+}
+
+/// Run a command on directories with a `pubspec.yaml`.
+Future<void> _runCommand<T>({
+  required Future<T> Function(String cwd) cmd,
+  required String cwd,
+  required bool recursive,
+}) async {
+  if (!recursive) {
+    final pubspec = File(p.join(cwd, 'pubspec.yaml'));
+    if (!pubspec.existsSync()) throw PubspecNotFound();
+
+    await cmd(cwd);
+    return;
+  }
 
-  /// Run a command on directories with a `pubspec.yaml`.
-  static Future<void> _runCommand({
-    required Future<ProcessResult> Function(String cwd) cmd,
-    required String cwd,
-    required bool recursive,
-  }) async {
-    if (!recursive) {
-      final pubspec = File(p.join(cwd, 'pubspec.yaml'));
-      if (!pubspec.existsSync()) throw PubspecNotFound();
+  final processes = _Cmd.runWhere(
+    run: (entity) => cmd(entity.parent.path),
+    where: _isPubspec,
+    cwd: cwd,
+  );
 
-      await cmd(cwd);
-      return;
-    }
+  if (processes.isEmpty) throw PubspecNotFound();
 
-    final processes = _Cmd.runWhere(
-      run: (entity) => cmd(entity.parent.path),
-      where: _isPubspec,
-      cwd: cwd,
-    );
+  for (final process in processes) {
+    await process;
+  }
+}
 
-    if (processes.isEmpty) throw PubspecNotFound();
+Future<void> _flutterTest({
+  String cwd = '.',
+  required void Function(String) stdout,
```

Comment:
Nit: reuse the typedef here

Reply:
Same question here about whether you feel this is necessary? I'd prefer to leave it inline unless it's a public API (related to https://dart-lang.github.io/linter/lints/avoid_private_typedef_functions.html).

Reply:
Interesting lint rule, never thought about it that way, makes sense! Let's keep online, we can ignore all the similar comments

Reply:
The only reason I slightly prefer to leave it as is is because let's say we create a typedef like:

```dart
typedef ArgumentCallback<T> = void Function(T argument);
```

Then the new signature will look like:

```dart
Future<void> _flutterTest({
  String cwd = '.',
  required ArgumentCallback<String>? stdout,
  required ArgumentCallback<String>? stderr,
}) {...}
```

I don't really see the benefit of the typedef in this case because as a developer, I have to look at what `ArgumentCallback` means above and, from a readability standpoint, I personally prefer seeing the type inline when it's simple like this especially if it's used in a private method.

Let me know what you think 👍

Reply:
Got it, I think it makes sense, I kind find the inline function type a little bit confusing on dart, hence why I prefer to use typedef, and as a bonus we DRY.

But I agree with you, being a private method, we have have freedom to leave like this, and it is easier to spot the require method contract. Lets keep like that 👍